src/client/content_cao.cpp

```diff
@@ -1832,10 +1866,42 @@ void GenericCAO::processMessage(const std::string &data)
 		updateAnimationSpeed();
 	} else if (cmd == AO_CMD_SET_BONE_POSITION) {
 		std::string bone = deSerializeString16(is);
-		v3f position = readV3F32(is);
-		v3f rotation = readV3F32(is);
-		m_bone_position[bone] = core::vector2d<v3f>(position, rotation);
-
+		BoneOverride *previous = m_bone_override[bone];
+		BoneOverride *override = new BoneOverride();
+		override->position.vector = readV3F32(is);
+		override->rotation.next = core::quaternion(readV3F32(is) * core::DEGTORAD);
+		if (is.eof()) {
+			override->position.absolute = true;
+			override->rotation.absolute = true;
+		} else {
+			override->scale.vector = readV3F32(is);
+			override->position.interpolation_duration = readF32(is);
+			override->rotation.interpolation_duration = readF32(is);
+			override->scale.interpolation_duration = readF32(is);
+			u8 absoluteFlag = readU8(is);
+			if (absoluteFlag == 0
+					&& override->position.vector == v3f(0.0f, 0.0f, 0.0f)
+					&& override->rotation.next == core::quaternion()
+					&& override->scale.vector == v3f(1.0f, 1.0f, 1.0f)) {
+				m_bone_override.erase(bone); // identity override, remove
+			} else {
+				override->position.absolute = (absoluteFlag & 1) > 0;
+				override->rotation.absolute = (absoluteFlag & 2) > 0;
+				override->scale.absolute = (absoluteFlag & 4) > 0;
+				if (previous) {
+					override->position.previous = previous->position.vector;
+					override->rotation.previous = previous->rotation.next;
+					override->scale.previous = previous->scale.vector;
+				} else {
+					// disable interpolation
+					override->position.interpolation_duration = 0.0f;
+					override->rotation.interpolation_duration = 0.0f;
+					override->scale.interpolation_duration = 0.0f;
+				}
+				m_bone_override[bone] = override;
+			}
+		}
+		if (previous) delete previous;
```

Comment:
This is dangerous. Map assignments and object freeing should happen in the same scope to avoid dangling pointers caused by logic errors.

Generally, I would prefer to not use pointers in this case. A sequential memcpy on this struct is quick to execute during runtime.

Reply:
And indeed the `if` branch is lacking a map assignment! Will use a boolean var to keep both in the same scope.